### ISSUE TITLE
[2017.7.2] ESXi Proxy Minion errors

### ISSUE THREAD

**Issue by danner26** [saltstack/salt#44065]
### Description of Issue/Question
When running the ESXi Proxy Minion as described in the documentation, it errors out with: 
The Salt ProxyMinion is shutdown.
Proxymodule esxi is missing an init() or a shutdown() or both. Check your proxymodule.  Salt-proxy aborted.

### Setup
All files are setup exactly as the documentation sas, but for clarity:
/etc/salt/proxy
master: salt.domain.com

/srv/pillar/top.sls
base:
    'esxi':
      - esxi

/srv/pillar/esxi.sls
proxy:
  proxytype: esxi
  host: 192.168.1.10
  username: root
  passwords:
    - <removed password>

### Versions Report
[root@salt ~]# salt --versions-report
Salt Version:
           Salt: 2017.7.2

Dependency Versions:
           cffi: 1.11.2
       cherrypy: unknown
       dateutil: Not Installed
      docker-py: 1.10.6
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.7
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.2.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core

Attached is the debug output for the proxy command.
[debug.txt](https://github.com/saltstack/salt/files/1379432/debug.txt)

**Comment by garethgreenaway** [comment]
@danner26 Thanks for the report.  Looking at the debug.txt that you provided it appears that the proxy module is not loaded because the vsphere execution module is not loaded.  Can you verify that the pyVmomi python library and the esxcli binary is installed?

**Comment by danner26** [comment]
@garethgreenaway Hello, yes I did see that the module is not loading. I do have both of these packages installed..

[root@salt ~]# pip install pyvmomi
Requirement already satisfied: pyvmomi in /usr/lib/python2.7/site-packages
Requirement already satisfied: requests>=2.3.0 in /usr/lib/python2.7/site-packages (from pyvmomi)
Requirement already satisfied: six>=1.7.3 in /usr/lib/python2.7/site-packages (from pyvmomi)

[root@salt ~]# esxcli --version
Script 'esxcli' version: 6.5.0

**Comment by bharrison-it** [comment]
We're utilizing the same setup and experiencing the same issue with some platform differences; Details below:

**Versions**
```
root@na:/# salt --versions-report
Salt Version:
           Salt: 2017.7.2
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.2
      docker-py: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: 2.10
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.9 
         locale: UTF-8
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.9 

```
**Python Deps:**
pip install pyVmomi
Requirement already satisfied: pyVmomi in /usr/local/lib/python2.7/dist-packages
Requirement already satisfied: six>=1.7.3 in /usr/lib/python2.7/dist-packages (from pyVmomi)
Requirement already satisfied: requests>=2.3.0 in /usr/lib/python2.7/dist-packages (from pyVmomi)

**ESXCLI Version**
Script 'esxcli' version: 6.0.0

[debug.log](https://github.com/saltstack/salt/files/1425549/debug.log)

Likewise, I can see the successful logon in the ESXI host event log. 

Maybe we can pull the module from dev as a workaround?

**Comment by snw1968** [comment]
Is there any update on this as I have the same issue - and I am stuck !!!

**Comment by lomeroe** [comment]
Issue still exists in 2017.7.5

My limited troubleshooting has come up with:

vsphere.system_info is not available on my system, unless I comment out or remove the '@ignores_keywords' decorator (same goes for any function with that decorator in vsphere.py).  It seems to be the only module using that particular decorator. ping @skazi0 
```
@depends(HAS_PYVMOMI)
#ignores_kwargs('credstore')
def system_info(host, username, password, protocol=None, port=None):
```

Once vsphere.system_info is available:

The __salt__ dunder seems to be empty in the esxi proxy's __virtual__ function

setting __virtual__ to print __salt__, returns an empty dict:
```
[DEBUG   ] __salt__ == {}
```
Which makes __virtual__ return False, since it is checking for vsphere.system_info in __salt__.

Changing __virtual__ to just return True gets the minion to start, and test.ping and grains can also be returned from the minion (which makes no sense to me as the grains function uses the __salt__ dunder).

The __virtual__ function for the esxi proxy module is different in develop, so I may do a PR for a similar setup.  Curious of @skazi0 thoughts on the ignores_kwargs decorator.

**Comment by lomeroe** [comment]
this seems to be fixed in 2018.3 for me

**Comment by danner26** [comment]
I have abandoned saltstack, so I will not be able to confirm but if anyone else is, then feel free to close this!

**Comment by bharrison-it** [comment]
I have embraced saltstack, so I can take up the mantle and test this out.

**Comment by ironman627** [comment]
I thought I had this fixed, with an older version of pyVmomi (7.0.3) but I lost my salt-master, had to reinstall it, and now I seem to be in a catch-22 situation:
I either get: `salt.exceptions.InvalidConfigError: Additional properties are not allowed ('credstore' was unexpected)`

Or if I pull the credstore out and run the proxy-minion again, I get:
`salt.exceptions.VMwareConnectionError: Could not verify the SSL certificate. You can use verify_ssl: False if you do not want to verify the SSL certificate.`

If I put the verify_ssl: False into pillar, then I get:
`salt.exceptions.InvalidConfigError: Additional properties are not allowed ('verify_ssl' was unexpected)`

Current salt-master has the following:
```
Script 'esxcli' version: 6.0.0
[root@localhost pillar]# salt-pip list | grep -E 'pyvmomi|json'
jsonschema                4.25.1
jsonschema-specifications 2025.9.1
pyvmomi                   7.0.3

Salt Version:
          Salt: 3006.16

Python Version:
        Python: 3.10.18 (main, Sep  5 2025, 22:48:51) [GCC 11.2.0]

```

I have looked at the proxy module for [esxi.py](http://esxi.py/) and it documents “credstore”. Pillar file versions have looked like this:

```
proxy:
  proxytype: esxi
  host: 192.168.0.4
  credstore: /root/.vmware/credstore/vicredentials.xml
  #verify_ssl: False
  username: root
  passwords:
    - "mySuperSecretPassword"

```

```
proxy:
  proxytype: esxi
  host: 192.168.0.4
  #credstore: /root/.vmware/credstore/vicredentials.xml
  verify_ssl: False
  username: root
  passwords:
    - "mySuperSecretPassword"

```
Have also removed username/passwords altogether and just referenced “credstore”.

Vicredentials.xml file looks like:

```
<?xml version="1.0" encoding="UTF-8"?>
<vicredentials>
  <credential>
    <host>ESXi-1.local</host>
    <user>root</user>
    <password>mySuperSecretPassword</password>
  </credential>
  <credential>
    <host>vcenter.local</host>
    <user>administrator@vsphere.local</user>
    <password>mySuperSecretPassword</password>
  </credential>
</vicredentials>
```

Have also tried it with IP addresses for the ESX and vCenter hosts/VMs.
